### ISSUE TITLE
Tunnel hosts using function of previous hosts

### ISSUE THREAD

**Issue by MadsRC** [cmp-202/ssh2shell#33]
Source:
```
var SSH2Shell = require('ssh2shell');
var msg = {
  send: function( message ) {
    console.log(message);
  }
 }

var remoteTarget = {
  server: {
    host: "172.22.20.51",
    userName: "user",
    password: "PASSWORD"
  },
  idleTimeOut: 6000,
  debug: true,
  commands: [
    'show hostname',
  ],
  msg: msg,
  onCommandComplete: function processCommandOutput(command, response, sshObj) {
    if (command === 'show hostname') {
      this.emit("msg", 'N - Hostname: '+ response)
    }
  },
  onCommandTimeout: function logTimeout(command, response, sshObj, stream, connection) {
      this.emit("msg", 'N - Command timed out: ' + command + ' (response: ' + response + ')');
  },
  onEnd: function sshDone(sessionText, sshObj) {
    this.emit("msg", 'N - done sending commands, over ssh, to the Nexus');
  }
};
var proxy = {
  server:        {
    host:         "172.22.20.20",
    userName:     "root",
    password:     "PASSWORD",
  },
  debug: true,
  commands:      [
    "msg:Connected to jump host!",
    "hostname",
  ],
  msg: msg,
  onCommandComplete: function processCommandOutput(command, response, sshObj) {
    if (command === 'show hostname') {
      this.emit("msg", 'J - Hostname: ' + response);
    }
  },
  onCommandTimeout: function logTimeout(command, response, sshObj, stream, connection) {
      this.emit("msg", 'J - Command timed out: ' + command + ' (response: ' + response + ')');
  },
  onEnd: function sshDone(sessionText, sshObj) {
    this.emit("msg", 'J - done sending commands, over ssh, to the Jump Host');
  }
};

proxy.hosts = [remoteTarget];

var SSH = new SSH2Shell(proxy);
SSH.connect();
```

Output:
```
172.22.20.20: Connection.connect
Connected
172.22.20.20: Connection.ready
Ready
172.22.20.20: Connection.shell
172.22.20.20: Normal prompt detected
172.22.20.20: Note: Connected to jump host!
172.22.20.20: Next command: hostname
172.22.20.20: Normal prompt detected
172.22.20.20: Queued hosts for this host
[ { server: 
     { host: '172.22.20.51',
       userName: 'user',
       password: 'PASSWORD' },
    idleTimeOut: 6000,
    debug: true,
    commands: [ 'show hostname' ],
    msg: { send: [Function: send] },
    onCommandComplete: [Function: processCommandOutput],
    onCommandTimeout: [Function: logTimeout],
    onEnd: [Function: sshDone] } ]
172.22.20.20: SSH to 172.22.20.51
172.22.20.51: Next command: ssh -oStrictHostKeyChecking=no user@172.22.20.51
172.22.20.51: SSH password prompt
172.22.20.51: SSH normal prompt
172.22.20.51: Next command: show hostname
172.22.20.51: Normal prompt detected
J - Hostname: show hostname
NX-5548-CORE-1 
NX-5548-CORE-1# 
172.22.20.51: Queued exit commands: exit
172.22.20.51: Next command: exit
172.22.20.51: Normal prompt detected
J - done sending commands, over ssh, to the Jump Host

Previous host object:
{ server: 
   { host: '172.22.20.20',
     userName: 'root',
     password: 'PASSWORD',
     hashKey: '' },
  debug: true,
  commands: [],
  msg: { send: [Function: send] },
  onCommandComplete: [Function: processCommandOutput],
  onCommandTimeout: [Function: logTimeout],
  onEnd: [Function: sshDone],
  hosts: [],
  connectedMessage: 'Connected',
  readyMessage: 'Ready',
  closedMessage: 'Closed',
  verbose: false,
  standardPrompt: '>$%#',
  passwordPromt: ':',
  passphrasePromt: ':',
  enter: '\n',
  asciiFilter: '[^\r\n -~]',
  disableColorFilter: false,
  textColorFilter: '([{1}[0-9;]+m{1})',
  exitCommands: [],
  pwSent: false,
  sshAuth: false,
  sessionText: 'Connected to 172.22.20.20\nhostname\r\nbackupproxy.example.com\r\n[root@backupproxy ~]# ',
  idleTimer: 
   Timeout {
     _called: false,
     _idleTimeout: 5000,
     _idlePrev: 
      TimersList {
        _idleNext: [Circular],
        _idlePrev: [Circular],
        _timer: [Object],
        _unrefed: false,
        msecs: 5000 },
     _idleNext: 
      TimersList {
        _idleNext: [Circular],
        _idlePrev: [Circular],
        _timer: [Object],
        _unrefed: false,
        msecs: 5000 },
     _idleStart: 320,
     _onTimeout: [Function],
     _repeat: null } }
172.22.20.20: Reload previous host object
172.22.20.20: Exit command: Stream: close
172.22.20.20: Stream.finish
J - done sending commands, over ssh, to the Jump Host
172.22.20.20: Stream.close
172.22.20.20: Connection.close
Closed
```

I'm seeing that none of the functions on the `remoteTarget` host is being fired and that the functions on `proxy` is being used for `remoteTarget`.

The README states that each host should have it's own set of functions defined, thus indicating that the above output indicates a bug.

It can be worked around, by having all the logic in the `proxy` host, however this quickly becomes rather messy.

**Comment by MadsRC** [comment]
Oh, and the reason for the `hostname` command on the `proxy` host is that without it, this is the result:

```
172.22.20.20: Connection.connect
Connected
172.22.20.20: Connection.ready
Ready
172.22.20.20: Connection.shell
172.22.20.20: Normal prompt detected
172.22.20.20: Note: Connected to jump host!
172.22.20.20: Queued hosts for this host
[ { server: 
     { host: '172.22.20.51',
       userName: 'user',
       password: 'PASSWORD' },
    idleTimeOut: 6000,
    debug: true,
    commands: [ 'show hostname' ],
    msg: { send: [Function: send] },
    onCommandComplete: [Function: processCommandOutput],
    onCommandTimeout: [Function: logTimeout],
    onEnd: [Function: sshDone] } ]
172.22.20.20: SSH to 172.22.20.51
172.22.20.51: Class.error
Error: TypeError: this.command.indexOf is not a function TypeError: this.command.indexOf is not a function
    at SSH2Shell._processNextCommand (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:176:24)
    at SSH2Shell._processNextCommand (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:4:59)
    at SSH2Shell._nextHost (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:224:19)
    at SSH2Shell._nextHost (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:4:59)
    at SSH2Shell._runExit (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:242:21)
    at SSH2Shell._runExit (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:4:59)
    at SSH2Shell._processNextCommand (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:197:23)
    at SSH2Shell._processNextCommand (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:4:59)
    at SSH2Shell._processData (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:50:21)
    at SSH2Shell._processData (/home/mrc/repo/wwi/confbackup/node_modules/ssh2shell/lib/ssh2shell.js:4:59), Level: Data handling
172.22.20.51: Connection.close
Closed
172.22.20.51: Stream.finish
J - done sending commands, over ssh, to the Jump Host
172.22.20.51: Stream.close
J - Command timed out: false (response: )
```

**Comment by cmp-202** [comment]
Without looking at you routput I note that you have an extra `,` in you command array to both remoteTarget and proxy

**Comment by MadsRC** [comment]
The trailing `,`? Just tried removing it, with no effect to the output.

**Comment by MadsRC** [comment]
Sorry, wrong button 💃

**Comment by cmp-202** [comment]
Looking at your output both command are being processed on the correct host:
```
172.22.20.20: Normal prompt detected
172.22.20.20: Note: Connected to jump host!
172.22.20.20: Next command: hostname
172.22.20.20: Normal prompt detected
```
and
```
172.22.20.51: Next command: show hostname
172.22.20.51: Normal prompt detected
J - Hostname: show hostname
NX-5548-CORE-1 
NX-5548-CORE-1# 
```

**Comment by cmp-202** [comment]
try adding `this.emit("msg", 'Session Text:\n' + sessionText);` to the onEnd handler of your proxy to see the session text for the remote host and then the proxy.

**Comment by cmp-202** [comment]
what I am seeing is that you ronENd handel for the remote host that should output `done sending commands, over ssh, to the Nexus` didn't show up in the output.

**Comment by MadsRC** [comment]
Output:
```
172.22.20.20: Connection.connect
Connected
172.22.20.20: Connection.ready
Ready
172.22.20.20: Connection.shell
172.22.20.20: Normal prompt detected
172.22.20.20: Note: Connected to jump host!
172.22.20.20: Next command: hostname
172.22.20.20: Normal prompt detected
J - Hostname: hostname
backupproxy.example.com
[root@backupproxy ~]# 
172.22.20.20: Queued hosts for this host
[ { server: 
     { host: '172.22.20.51',
       userName: 'user',
       password: 'PASSWORD' },
    idleTimeOut: 6000,
    debug: true,
    commands: [ 'show hostname' ],
    msg: { send: [Function: send] },
    onCommandComplete: [Function: processCommandOutput],
    onCommandTimeout: [Function: logTimeout],
    onEnd: [Function: sshDone] } ]
172.22.20.20: SSH to 172.22.20.51
J - Hostname: 
172.22.20.51: Next command: ssh -oStrictHostKeyChecking=no user@172.22.20.51
172.22.20.51: SSH password prompt
172.22.20.51: SSH normal prompt
172.22.20.51: Next command: show hostname
172.22.20.51: Normal prompt detected
172.22.20.51: Queued exit commands: exit
172.22.20.51: Next command: exit
172.22.20.51: Normal prompt detected
J - done sending commands, over ssh, to the Jump Host
Session Text:
Connected to 172.22.20.51
show hostname
NX-5548-CORE-1 
NX-5548-CORE-1# 

Previous host object:
{ server: 
   { host: '172.22.20.20',
     userName: 'root',
     password: 'PASSWORD',
     hashKey: '' },
  debug: true,
  commands: [],
  msg: { send: [Function: send] },
  onCommandComplete: [Function: processCommandOutput],
  onCommandTimeout: [Function: logTimeout],
  onEnd: [Function: sshDone],
  hosts: [],
  connectedMessage: 'Connected',
  readyMessage: 'Ready',
  closedMessage: 'Closed',
  verbose: false,
  standardPrompt: '>$%#',
  passwordPromt: ':',
  passphrasePromt: ':',
  enter: '\n',
  asciiFilter: '[^\r\n -~]',
  disableColorFilter: false,
  textColorFilter: '([{1}[0-9;]+m{1})',
  exitCommands: [],
  pwSent: false,
  sshAuth: false,
  sessionText: 'Connected to 172.22.20.20\nhostname\r\nbackupproxy.example.com\r\n[root@backupproxy ~]# ',
  idleTimer: 
   Timeout {
     _called: false,
     _idleTimeout: 5000,
     _idlePrev: 
      TimersList {
        _idleNext: [Circular],
        _idlePrev: [Circular],
        _timer: [Object],
        _unrefed: false,
        msecs: 5000 },
     _idleNext: 
      TimersList {
        _idleNext: [Circular],
        _idlePrev: [Circular],
        _timer: [Object],
        _unrefed: false,
        msecs: 5000 },
     _idleStart: 302,
     _onTimeout: [Function],
     _repeat: null } }
172.22.20.20: Reload previous host object
172.22.20.20: Exit command: Stream: close
172.22.20.20: Stream.finish
J - done sending commands, over ssh, to the Jump Host
Session Text:
Connected to 172.22.20.20
hostname
backupproxy.example.com
[root@backupproxy ~]# 
172.22.20.20: Stream.close
172.22.20.20: Connection.close
Closed
```

The commands are executed correctly and all. I just wonder why proxy.onCommandComplete seems to be fired for commands on remoteTarget.commands instead of remoteTarget.onCommandComplete

**Comment by cmp-202** [comment]
Yes I think that is similar to the onEnd for the remote not firing. That is definitly a bug caused by the event handler being defined when the instance is created but if it isn't switched when a new host is loaded or a previous host is reloaded than only the first host event handlers will be firing. Good spotting that man!

**Comment by cmp-202** [comment]
Okay can you grab the lib/ssh2shell.js file from master and try it

**Comment by MadsRC** [comment]
source:
```
var SSH2Shell = require('ssh2shell');
var msg = {
  send: function( message ) {
    console.log(message);
  }
 }

var remoteTarget = {
  server: {
    host: "172.22.20.51",
    userName: "user",
    password: "PASSWORD"
  },
  idleTimeOut: 6000,
  commands: [
    'show hostname'
  ],
  msg: msg,
  onCommandComplete: function processCommandOutput(command, response, sshObj) {
    if (command === 'show hostname') {
      this.emit("msg", 'N - Hostname: '+ response)
    }
  },
  onCommandTimeout: function logTimeout(command, response, sshObj, stream, connection) {
      this.emit("msg", 'N - Command timed out: ' + command + ' (response: ' + response + ')');
  },
  onEnd: function sshDone(sessionText, sshObj) {
    this.emit("msg", 'N - done sending commands, over ssh, to the Nexus');
  }
};
var proxy = {
  server:        {
    host:         "172.22.20.20",
    userName:     "root",
    password:     "PASSWORD",
  },
  commands:      [
    "msg:Connected to jump host!",
    "hostname"
  ],
  msg: msg,
  onCommandComplete: function processCommandOutput(command, response, sshObj) {
    if (command === 'hostname') {
      this.emit("msg", 'J - Hostname: ' + response);
    }
    else if (command === 'show hostname') {
      this.emit("msg", 'J - Hostname: '+ response)
    }
  },
  onCommandTimeout: function logTimeout(command, response, sshObj, stream, connection) {
      this.emit("msg", 'J - Command timed out: ' + command + ' (response: ' + response + ')');
  },
  onEnd: function sshDone(sessionText, sshObj) {
    this.emit("msg", 'J - done sending commands, over ssh, to the Jump Host');
  }
};

proxy.hosts = [remoteTarget];

var SSH = new SSH2Shell(proxy);
SSH.connect();
```

output:
```
Connected
Ready
172.22.20.20: Note: Connected to jump host!
J - Hostname: hostname
backupproxy.example.com
[root@backupproxy ~]# 
J - Hostname: 
J - Hostname: 
J - Hostname: show hostname
NX-5548-CORE-1 
NX-5548-CORE-1# 
J - Hostname: show hostname
NX-5548-CORE-1 
NX-5548-CORE-1# 
N - Hostname: show hostname
NX-5548-CORE-1 
NX-5548-CORE-1# 
N - Hostname: show hostname
NX-5548-CORE-1 
NX-5548-CORE-1# 
J - done sending commands, over ssh, to the Jump Host
J - done sending commands, over ssh, to the Jump Host
N - done sending commands, over ssh, to the Nexus
N - done sending commands, over ssh, to the Nexus
J - done sending commands, over ssh, to the Jump Host
J - done sending commands, over ssh, to the Jump Host
N - done sending commands, over ssh, to the Nexus
N - done sending commands, over ssh, to the Nexus
Closed
Closed
```

There's quite a few "show hostname" there?
The expected output would be no "```
J - Hostname: show hostname
NX-5548-CORE-1 
NX-5548-CORE-1#
```"
but only "```
N - Hostname: show hostname
NX-5548-CORE-1 
NX-5548-CORE-1#
```"

**Comment by cmp-202** [comment]
hmmm bugger ok I will have to trace that through and fire up my VM's

**Comment by cmp-202** [comment]
I know what it is doing (grr) instead of replacing the event handler it is adding another one. The old method just called the host event handlers from within the event function but it meant you couldn't access this keywork and therefore the instance. The change I made to enable that has broken the multiple hosts event definitions.

**Comment by cmp-202** [comment]
Give that a whurl

**Comment by MadsRC** [comment]
That solved my problem!

Would you happen to have paypal or somewhere else I can throw a tip your way?
You've been awesome, and this project is awesome!

**Comment by cmp-202** [comment]
I am setting class veriables to each host handler/default handler function then adding the listerner to the emitter using each function when loading the defaults. Before switching between hosts I remove the event handlers with those functions, load the new host and reload the defaults which loads the events with new functions if there are any. I replaced some of the default definitions to only set a value if one doesn't exist because the sessionText for the first host was getting wiped.

**Comment by cmp-202** [comment]
All good all part of the node community ethos. The results from this work is a much better ssh2shell package. Any input on those other two open issues would be appreciated if you have any.

**Comment by MadsRC** [comment]
I'll leave some input in the issues :D

Seem to be a minor issue with the idleTimeout on the proxy being fired during processing of the remoteTarget's commands. Can be worked around pretty easily by setting the proxy's timeout sufficiently high. Want me to create a new issue with some example and some code?

**Comment by cmp-202** [comment]
Yes please. I published these changes to npm v1.5.10.
